pootle/apps/pootle_translationproject/views.py

```diff
@@ -43,120 +57,323 @@ def admin_permissions(request, translation_project):
-    """Displays a list of units with filters applied."""
-    ctx = get_export_view_context(request)
-    ctx.update({
-        'source_language': translation_project.project.source_language,
-        'language': translation_project.language,
-        'project': translation_project.project,
-    })
-
-    return render(request, 'editor/export_view.html', ctx)
+def redirect_to_tp_on_404(f):
+
+    @functools.wraps(f)
+    def method_wrapper(self, request, *args, **kwargs):
+        User = get_user_model()
+        request.profile = User.get(self.request.user)
+        try:
+            request.permissions = get_matching_permissions(
+                request.profile,
+                self.permission_context) or []
+        except Http404 as e:
+            if (kwargs["dir_path"] or kwargs.get("filename", None)):
+                try:
+                    TranslationProject.objects.get(
+                        project__code=kwargs["project_code"],
+                        language__code=kwargs["language_code"])
+                    # the TP exists so redirect to it
+                    return redirect(
+                        reverse(
+                            'pootle-tp-browse',
+                            kwargs={
+                                k: v
+                                for k, v
+                                in kwargs.items()
+                                if k in [
+                                    "language_code",
+                                    "project_code"]}))
+                except TranslationProject.DoesNotExist:
+                    pass
+
+            # if we get here - the TP does not exist
+            user_choice = self.request.COOKIES.get(
+                'user-choice', None)
+            if user_choice:
+                url = None
+                if user_choice == 'language':
+                    url = reverse(
+                        'pootle-language-browse',
+                        args=[kwargs["language_code"]])
+                elif user_choice == "project":
+                    url = reverse(
+                        'pootle-project-browse',
+                        args=[kwargs["project_code"], '', ''])
+                if url:
+                    response = redirect(url)
+                    response.delete_cookie('user-choice')
+                    return response
+            raise e
+        return f(self, request, *args, **kwargs)
+    return method_wrapper
+
+
+class TPMixin(object):
+
+    @redirect_to_tp_on_404
+    @set_permissions
+    @requires_permission("view")
+    def dispatch(self, request, *args, **kwargs):
+        # get funky with the request 8/
+        return super(TPMixin, self).dispatch(request, *args, **kwargs)
+
+    @property
+    def ctx_path(self):
+        return self.tp.pootle_path
+
+    @property
+    def resource_path(self):
+        return self.object.pootle_path.replace(self.ctx_path, "")
```

Comment:
that was confusing because of class name. I didn't see that you inherit store and directory mixins from this.